addons/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/converters/ConverterFactory.java

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.ihc.internal.converters;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.types.UpDownType;
+import org.eclipse.smarthome.core.types.Type;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSBooleanValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSDateValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSEnumValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSFloatingPointValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSIntegerValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSResourceValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSTimeValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSTimerValue;
+import org.openhab.binding.ihc.internal.ws.resourcevalues.WSWeekdayValue;
+
+/**
+ * IHC / ELKO <-> openHAB data type converter factory.
+ *
+ * @author Pauli Anttila - Initial contribution
+ */
+public enum ConverterFactory {
+    INSTANCE;
+
+    private static final String ITEM_TYPE_NUMBER = "Number";
+    private static final String ITEM_TYPE_SWITCH = "Switch";
+    private static final String ITEM_TYPE_CONTACT = "Contact";
+    private static final String ITEM_TYPE_DIMMER = "Dimmer";
+    private static final String ITEM_TYPE_DATETIME = "DateTime";
+    private static final String ITEM_TYPE_STRING = "String";
+    private static final String ITEM_TYPE_ROLLERSHUTTER = "Rollershutter";
+
+    private class Key {
+        Class<? extends WSResourceValue> ihcType;
+        Class<? extends Type> openhabType;
+
+        Key(Class<? extends WSResourceValue> ihcType, Class<? extends Type> openhabType) {
+            this.ihcType = ihcType;
+            this.openhabType = openhabType;
+        }
+
+        @Override
+        public int hashCode() {
+            return new String(ihcType.getClass().toString() + openhabType.getClass().toString()).hashCode();
```

Comment:
Is there a change for collision, at least in theory?

5+6=11
7+4=11

Reply:
Yes, but that is completely fine for hashcodes, in some cases it could have minor performance impact, but I believe the performance impact of the old implementation is bigger. 

Some background, if you use classes with hash in the name with these classes, for example the `HashMap` buckets are created using the `hashcode` inside a bucket every item is compared with `equals`. So the implementation 42 is (of course) also correct, but it will cause all items to be in single bucket and make the performance of the HashMap go from `O(Log(N))` to `O(N)` if I recall that piece of theory correctly :-).

Because of that collaboration there a lot of rules regarding to a (completely) correct implementation of the `equals`/`hashcode`, a former college of my made a great tool to test it so you include it in your tests as well :-)

https://jqno.nl/equalsverifier/

Reply:
Ok...done